Model/ExtendedCreditMemoFactory.php

```diff
@@ -0,0 +1,101 @@
+     */
+    private $refundableItemsFilter;
+
+    /**
+     * @var CreditmemoFactory
+     */
+    private $creditmemoFactory;
+
+    /**
+     * @param RefundableItemsFilter $refundableItemsFilter
+     * @param CreditmemoFactory $creditmemoFactory
+     */
+    public function __construct(
+        RefundableItemsFilter $refundableItemsFilter,
+        CreditmemoFactory $creditmemoFactory
+    ) {
+        $this->refundableItemsFilter = $refundableItemsFilter;
+        $this->creditmemoFactory = $creditmemoFactory;
+    }
+
+    /**
+     * Create a credit memo based on the order
+     * @param OrderInterface $order
+     * @param CreditmemoInterface $creditmemo
+     * @return Creditmemo
+     * @throws SnowCreditMemoException
+     * @author Alexander Wanyoike <amw@amp.co>
+     */
+    public function create(OrderInterface $order, CreditmemoInterface $creditmemo)
+    {
+        $refundableItems = $this->refundableItemsFilter->filter($order, $creditmemo, $this->hasAdjustments($creditmemo));
+        if (empty($refundableItems) && !$this->hasAdjustments($creditmemo)) {
+            throw new SnowCreditMemoException(__('No items available to refund'));
+        }
+
+        if ($orderInvoice = $this->getLatestPaidInvoiceForOrder($order)) {
+            return $this->creditmemoFactory->createByInvoice($orderInvoice, [
+                'qtys' => $refundableItems,
+                'shipping_amount' => $creditmemo->getShippingAmount(),
+                'adjustment_positive' => $creditmemo->getBaseAdjustmentPositive(),
+                'adjustment_negative' => $creditmemo->getBaseAdjustmentNegative()
+            ]);
+        }
+
+        return $this->creditmemoFactory->createByOrder($order, [
+            'qtys' => $refundableItems,
+            'shipping_amount' => $creditmemo->getShippingAmount(),
+            'adjustment_positive' => $creditmemo->getBaseAdjustmentPositive(),
+            'adjustment_negative' => $creditmemo->getBaseAdjustmentNegative()
+        ]);
+    }
+
+    /**
+     * @param CreditmemoInterface $creditmemo
+     * @return bool
+     * @author Alexander Wanyoike <amw@amp.co>
+     */
+    private function hasAdjustments(CreditmemoInterface $creditmemo)
+    {
+        return $creditmemo->getBaseAdjustmentNegative() !== null ||
+            $creditmemo->getBaseAdjustmentPositive() !== null ||
+            $creditmemo->getBaseAdjustment() !== null;
+    }
+
+    /**
+     * Get latest invoice for order
+     *
+     * @param OrderInterface $order
+     * @return Invoice|null
+     */
+    private function getLatestPaidInvoiceForOrder(OrderInterface $order)
+    {
+        /** @var Invoice $latestInvoice */
+        $latestInvoice = $order->getInvoiceCollection()
+            ->addAttributeToFilter('state', ['eq' => Invoice::STATE_PAID])
+            ->setPageSize(1)
+            ->setCurPage(1)
+            ->getLastItem();
+
+        return $latestInvoice->getId() ? $latestInvoice : null;
```

Comment:
Might be best doing:
```php
$latestInvoice->getId() !== null ? $latestInvoice : null;
```
Just in case `getId` returns `0`

Reply:
Yeah ill make that change and see If add an integration test.